### ISSUE TITLE
lxcfs with systemd containers cause containers hang after a while

### ISSUE THREAD

**Issue by fajarnugraha** [lxc/lxcfs#17]
- Create vivid container with systemd
```
# lxc-create -n c1 -- -d ubuntu -r vivid -a amd64
# lxc-attach -n c1 apt-get update
# lxc-attach -n c1 apt-get install systemd-sysv
# lxc-stop -n c1
```

- Start-stop container repeatedly
```
# n=0;while true;do n=$((n+1));echo "$(date) -- Try #$n";sleep 1;lxc-start -n c1 && sleep 10 && lxc-attach -n c1 -- poweroff && lxc-console -n c1 -t console;done
```

Eventually the container will hang. This is the result of my latest test
```
Thu Mar 26 14:50:03 WIB 2015 -- Try #12
Failed to start poweroff.target: Connection timed out

Connected to tty 0
Type <Ctrl+a q> to exit the console, <Ctrl+a Ctrl+a> to enter Ctrl+a itself
```

At this time another ``lxc-attach -n c1`` command from the host will hang.

I run this test with lxcfs started using ``strace -f /usr/bin/lxcfs -s -f -o allow_other /var/lib/lxcfs 2>&1 | tee /data/lxcfs.log``, and last lines of output are
```
clone(child_stack=0, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7fe18bdcfa10) = 573
Process 573 attached
[pid  3156] select(5, [4], NULL, NULL, {2, 0} <unfinished ...>
[pid   573] set_robust_list(0x7fe18bdcfa20, 24) = 0
[pid   573] open("/proc/31766/ns/pid", O_RDONLY) = 6
[pid   573] setns(6, 0)                 = 0
[pid   573] close(6)                    = 0
[pid   573] clone(Process 574 attached
child_stack=0, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7fe18bdcfa10) = 574
[pid   573] wait4(574,  <unfinished ...>
[pid   574] futex(0x7fe18aee2760, FUTEX_WAIT_PRIVATE, 2, NULL <unfinished ...>
[pid  3156] <... select resumed> )      = 0 (Timeout)
[pid  3156] recvfrom(4, 0x7fffcc209c4f, 1, 64, 0, 0) = -1 EAGAIN (Resource temporarily unavailable)
[pid  3156] write(2, "send_creds: Error getting reply "..., 60send_creds: Error getting reply from server over socketpair
) = 60
[pid  3156] kill(573, SIGTERM)          = 0
[pid   573] <... wait4 resumed> 0x7fffcc209c94, 0, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
[pid  3156] wait4(573,  <unfinished ...>
[pid   573] --- SIGTERM {si_signo=SIGTERM, si_code=SI_USER, si_pid=3156, si_uid=0} ---
[pid   573] rt_sigreturn()              = -1 EINTR (Interrupted system call)
[pid   573] wait4(574, 
```

``573`` and ``574`` are lxcfs process. From the output above it looks like ``573`` is waiting for ``574``, while ``574`` is stuck on 
```[pid   574] futex(0x7fe18aee2760, FUTEX_WAIT_PRIVATE, 2, NULL <unfinished ...>```

When I run ``strace -f lxc-attach -n c1 2>&1 | tee /data/lxc-attach.log``, the last lines of output are
```
[pid   752] getuid()                    = 0
[pid   752] getgid()                    = 0
[pid   752] geteuid()                   = 0
[pid   752] getegid()                   = 0
[pid   752] rt_sigprocmask(SIG_BLOCK, NULL, [], 8) = 0
[pid   752] ioctl(0, SNDCTL_TMR_TIMEBASE or SNDRV_TIMER_IOCTL_NEXT_DEVICE or TCGETS, {B38400 opost isig icanon echo ...}) = 0
[pid   752] ioctl(2, SNDCTL_TMR_TIMEBASE or SNDRV_TIMER_IOCTL_NEXT_DEVICE or TCGETS, 0x7fffa8acde60) = -1 ENOTTY (Inappropriate ioctl for device)
[pid   752] brk(0x1f78000)              = 0x1f78000
[pid   752] open("/proc/meminfo", O_RDONLY|O_CLOEXEC
```

So what I gather so far:
- lxcfs stuck on some futex (don't know what exactly)
- any another process which access lxcfs (including lxc-attach, which in the case above simply reads ``/proc/meminfo`` will hang

**Comment by hallyn** [comment]
Thanks.  I've pushed some commits to address part of the problem, but there are still hangs - which turn out to be due to https://sourceware.org/bugzilla/show_bug.cgi?id=15392.   This causes fork() to actually hang on an assert.  I'll push the final fix later today (to kill and retry when fork hangs)

**Comment by hallyn** [comment]
this should be fixed in 0.7
